fv3config/_asset_list.py

```
@@ -0,0 +1,245 @@
+        )
+
+
+def write_asset_list(asset_list, target_directory):
+    """Loop over list of assets and write them all"""
+    for asset in asset_list:
+        write_asset(asset, target_directory)
+
+
+def check_asset_valid(asset):
+    """Check asset has all of its required keys"""
+    required_asset_properties = ['source_location', 'source_name', 'target_location',
+                                 'target_name', 'copy_method']
+    for required_asset_property in required_asset_properties:
+        if required_asset_property not in asset:
+            raise ConfigError(f'Assets must have a {required_asset_property}')
+
+
+def config_to_asset_list(config):
+    """Convert a configuration dictionary to an asset list. The asset list
+    will contain all files for the run directory except the namelist."""
+    asset_list = get_initial_conditions_asset_list(config)
+    asset_list += get_base_forcing_asset_list(config)
+    asset_list += get_orographic_forcing_asset_list(config)
+    asset_list.append(get_field_table_asset(config))
+    asset_list.append(get_diag_table_asset(config))
+    asset_list.append(get_data_table_asset(config))
+    if 'patch_files' in config:
+        if is_dict_or_list(config['patch_files']):
+            asset_list += ensure_is_list(config['patch_files'])
+        else:
+            raise ConfigError('patch_files item in config dictionary must be a dict or list')
+    return asset_list
+
+
+def link_file(source_item, target_item):
+    if os.path.exists(target_item):
+        os.remove(target_item)
+    os.link(source_item, target_item)
+
+
+def copy_file(source_path, target_path):
+    if is_gsbucket_url(source_path):
+        if gsutil_is_installed():
+            check_call(['gsutil', 'cp', source_path, target_path])
+        else:
+            logging.warning(f'Optional dependency gsutil not found. File {source_path} will not be copied to {target_path}')
```

Please raise an exception instead of a warning. I think this counts as a catastrophic failure.

---

Done. Also now raise the exception before trying to generate an asset_list from a gcs url if gsutil is not installed.